Devcontainer Setup v2

MareStare opened the pull request:
The first approach was https://github.com/philomena-dev/philomena/pull/528, back when it was my first time setting up a devcontainer. Not much has changed since then, except that I'd recommend using as less bind mounts as possible since every bind mount poses a requirement on the host.

This time, there is no "docker exec from host" support, and no `zsh` (the default `sh` is used). So the setup is quite a bunch simpler.

Meow commented:
Is it deliberate that the Dockerfile for the dev container lacks postgres client? It's required to dump and load database schema, see app/Dockerfile for reference

liamwhite commented:
pg_dump is needed to run mix ecto.migrate

MareStare commented:
I mean the devcontainer doesn't even have access to the DB, because Postgres DB port is not exposed.

Meow commented:
Okay, then I approve of this PR.

liamwhite commented:
Okay I caught it, it's from vscode forkbombing node processes
<img width="3818" height="1696" alt="Screenshot From 2025-11-09 10-53-46" src="https://github.com/user-attachments/assets/4b694dd1-b758-469f-8806-44d6ecd75909" />

It looks like this issue: https://github.com/microsoft/vscode/issues/199588

Veetaha commented:
I see, I'm using a remote SSH server to run this. Haven't tried on WSL, but I remember trying it long ago on WSL, and it worked before.

liamwhite commented:
The relevant code appears to be from the (distribution-only, no source) extension ms-vscode-remote with this string template:
```js
,s=await e({cmd:t,args:["-e",`// Connection forwarding to ${"port"in r?`${r.host}:${r.port}`:r.socketPath} on remote host.
const net = require('net');
const fs = require('fs');
process.stdin.pause();
const client = net.createConnection(${i}, () => {
        console.error('Connection established');
        client.pipe(process.stdout);
        process.stdin.pipe(client);
});
client.on('close', function (hadError) {
        console.error(hadError ? 'Remote close with error' : 'Remote close');
        process.exit(hadError ? 1 : 0);
});
client.on('error', function (err) {
        process.stderr.write(err && (err.stack || err.message) || String(err));
});
process.stdin.on('close', function (hadError) {
        console.error(hadError ? 'Remote stdin close with error' : 'Remote stdin close');
        process.exit(hadError ? 1 : 0);
});
process.on('uncaughtException', function (err) {
        fs.writeSync(process.stderr.fd, \`Uncaught Exception: \${String(err && (err.stack || err.message) || err)}\\n\`);
});
`
```
so it has something to do with the forwarding since the forkbomb only occurs once I try to connect in the browser

Veetaha commented:
I can remove forwarding configs. Vscode configures forwarding automatically anyway if you are using remote SSH, or one can manually forward ports too

Veetaha commented:
By "you" I rather meant "anyone"

MareStare commented on config/config.exs:
```diff
@@ -40,7 +40,9 @@ config :swoosh, :api_client, false
-  mode: :release
+  mode: :release,
+  # Never link the C runtime statically, because we are building a `dylib`
+  env: [{"RUSTFLAGS", "-C target-feature=-crt-static"}]
```
Not sure, I hope not

MareStare commented:
Per my intent, the devcontainer is separate from the app container. The app container is still part of the `docker-compose.yml` and the flow is basically the same, i.e. using `philomena up/down` (which uses `docker compose up/down` internally). The env var such as `ANONYMOUS_NAME_SALT` is part of the app container's docker-compose definition, for example, and for this reason it's not set for the devcontainer itself.

If you'd like to be able to run the app directly from the devcontainer itself, then it should be removed from the `docker-compose.yml` stack, and the env variables configs for it should be part of the `devcontainer.json`, which complicates things, because now you need to connect the `docker-compose.yml` settings and the `devcontainer.json` settings of the container.

MareStare commented:
Technically there is [docker-compose support](https://code.visualstudio.com/docs/devcontainers/create-dev-container#_use-docker-compose) in devcontainer. It'll bring up the stack when you open it in IDE, and you may point it to a specific service in the stack you want to develop in. I guess this is also an option. WDYT? 

This, however, limits the developer's ability to control the lifecycle of the docker-compose stack

liamwhite commented:
That's definitely way more in line with what I had envisioned--hopefully it ends up being close to what would actually be good for a development workflow :sweat_smile: 

BTW if it turns out alpine is too much of a pain, it's probably ok to switch to the debian slim images which use glibc

MareStare commented:
Finished updating the setup to make devcontainer use the `docker-compose.yml`. When you open the devcontainer it basically runs a `docker compose up` for you and attaches to the `app` service. Updated the docs a little bit, and updated the `philomena` scripts such that they make sure not to shutdown the `app` service of the stack under any condition.

Also, added an env var `DEVCONTAINER=0/1` that identifies if the process is running in devcontainer. There are a couple of `if`s with conditonal logic based on this env var to support running the stack outside of the devcontainer too, just in case.

liamwhite commented on README.md:
```diff
@@ -29,17 +29,11 @@ Once the application has started, navigate to http://localhost:8080 and login wi
+The devcontainer is configured with the `docker-compose.yml`. The IDE attaches to the `app` service of the stack. You can use `docker compose` to manage the stack but be careful not to shutdown the `app` service. Use `philomena` commands instead that make sure the `app` service is always running.
 
-If you are using VSCode, you are encouraged to install the recommended extensions that VSCode should automatically suggest to you based on `.vscode/extensions.json` file in this repo.
+If you are developing on a remote SSH host, then make sure to configure port forwarding for `web:80` (yes, use the `host:port` notation) to be able to access the application in your browser.
```
Did you mean :8080?
